File: examples/chapter04/e03_gram_matrix/rust/src/main.rs
use burn::backend::NdArray;
use burn::tensor::Tensor;

type B = NdArray;

fn main() {
    let dev = Default::default();

    // design matrix X: [n_samples, n_features]
    let x = Tensor::<B, 2>::from_floats([[1.0, 1.0], [1.0, 2.0], [1.0, 3.0]], &dev);
    let y = Tensor::<B, 2>::from_floats([[6.0], [0.0], [0.0]], &dev);

    let xt = x.clone().transpose();
    let gram = xt.clone().matmul(x.clone()); // X^T X
    let xty = xt.matmul(y.clone()); // X^T y
    // let w = linalg::solve(gram, xty); // <-- needs #1538

    println!("X^T X =\n{}", gram.to_data());
    println!("X^T y =\n{}", xty.to_data());
}

File: examples/chapter04/e03_gram_matrix/python/main.py
import torch
X = torch.tensor([[1., 1.],
                  [1., 2.],
                  [1., 3.]])       # [3 samples, 2 features]

y = torch.tensor([[6.], [0.], [0.]])

gram = X.T @ X                     # [[3., 6.], [6., 14.]]
xty  = X.T @ y                     # [[6.], [6.]]

w = torch.linalg.solve(gram, xty)  # the closed-form fit

print(f"X^T X =\n{gram}")
print(f"\nX^T y =\n{xty}")
print(f"\nw =\n{w}")
